src/servers/src/http/handler.rs

```diff
@@ -117,17 +123,49 @@ pub async fn sql(
         Ok(outputs) => outputs,
     };
 
-    let resp = match format {
+    let mut resp = match format {
         ResponseFormat::Arrow => ArrowResponse::from_output(outputs).await,
         ResponseFormat::Csv => CsvResponse::from_output(outputs).await,
         ResponseFormat::Table => TableResponse::from_output(outputs).await,
         ResponseFormat::GreptimedbV1 => GreptimedbV1Response::from_output(outputs).await,
         ResponseFormat::InfluxdbV1 => InfluxdbV1Response::from_output(outputs, epoch).await,
     };
 
+    if let Some(RequestSource::Dashboard) = source {
+        resp = from_dashboard(resp, query_ctx).await;
+    }
     resp.with_execution_time(start.elapsed().as_millis() as u64)
 }
 
+pub async fn from_dashboard(resp: HttpResponse, ctx: QueryContextRef) -> HttpResponse {
+    let HttpResponse::GreptimedbV1(response) = resp else {
+        return resp;
+    };
+    let GreptimedbV1Response {
+        output,
+        execution_time_ms,
+        resp_metrics,
+    } = response;
+    let query_limit = ctx.configuration_parameter().dashboard_query_limit();
+
+    let mut outputs = Vec::with_capacity(output.len());
+    for query_result in output {
+        let GreptimeQueryOutput::Records(mut records) = query_result else {
+            outputs.push(query_result);
+            continue;
+        };
+        if records.rows.len() > query_limit {
+            records.rows.resize(query_limit, vec![]);
+        }
```

Comment:
We should truncate the result instead of resizing it.
```suggestion
        records.rows.truncate(query_limit);
```